### ISSUE TITLE
Remove the limitation that Nyquist tool plugins cannot invoke other Nyquist plugins

### ISSUE THREAD

**Issue by werame** [audacity/audacity#2465]
**Is your feature request related to a problem? Please describe.**
A Nyquist plugin cannon presently call other Nyquist plugins. This creates numerous usability conundrums when scripting because Nyquist effect/generate plugins cannot call `aud-do` style commands. So if you write a tool plugin (which can call  `aud-do`) but that tries to call Nyquist fx plugins that way, these fail silently. 

**Describe the solution you'd like**
At least tool plugins should be able to call Nyquist effect plugins, successfully, as this is the prime use case, although other combinations should be considered if they can easily address in the same fix.

**Additional context**
This is pretty hard to solve because the NyquistEffect class gets instantiated to a singleton object that is non-reentrant, meaning a Nyquist plugin needs to fully finish before another can start, so that the state of this object is reset between runs. Probably what needs to be done is to at least have different sub-classes (and thus different singleton objects) for Nyquist tool vs fx plugins, so they have separate state. That way a tool plugin should be able to call fx plugins.

**Comment by Paul-Licameli** [comment]
This is not at all an easy issue.

My old colleagues won't like what I say, but the interface between Audacity and Nyquist in Nyquist.cpp is a big sprawl of confusing code, it was not well designed for reentrancy, and it has accumulated too many strange little hacks over the years.

It is done only in the special dynamic context of an effect-in-progress, which makes it not really appropriate for execution of tools using `aud-do`, but it was bent to that purpose.

The alternative scripting system, macros, which grew out of more limited-purpose effect chains, isn't robust enough to my liking either.  There is no clarity in the code about how the undo/redo stack should be pushed and popped, or where sqlite transactions should begin, end, or roll back.

All this leaves a lot to be cleaned up.  I want to do this cleanup.  But I am as always heavily occupied with dozens of other cleanups.

**Comment by Paul-Licameli** [comment]
Also problematic is how to cope with C++ exceptions that might want to escape macro or menu commands.  I put work into a better system of exception handling so these can propagate up to a handler in the main event loop -- if all the levels of stack in between are C++.  But the Nyquist interpreter is written in C, and uses setjmp and longjmp for other purposes.  Giving Nyquist a foreign function interface that might reenter C++ world, but can catch C++ exceptions, unwind the C stack levels (and the Lisp call stack levels implemented thereby), then rethrow the exception into C++ again -- well, all of that would require care.

**Comment by Paul-Licameli** [comment]
Still an urealized fantasy of mine is to give more nerdy Audacity users a REPL window -- a kind of command-line console interpreting Lisp in a loop, doing more than the Nyquist prompt does.

It would be tons of fun to figure out, but the business priority of it is dubitable.

**Comment by werame** [comment]
> accumulated too many strange little hacks over the years.

Indeed. For `;type tool` most of the fields in Nyquist.h [that deal with buffers and tracks](https://github.com/audacity/audacity/blob/master/src/effects/nyquist/Nyquist.h#L257) are pointless as Nyquist macros don't get those. And most of the state reset troubles come form those buffer/track fields. Which is why I'm thinking that making tools be able to call Ny effects by moving the tools to a separate class, or even just instantiating a new NyquistEffect object for them (as a quick hack) is easier than solving the re-entrance issue more broadly. 

> Giving Nyquist a foreign function interface that might reenter C++ world, but can catch C++ exceptions, unwind the C stack levels (and the Lisp call stack levels implemented thereby), then rethrow the exception into C++ again -- well, all of that would require care.

Doesn't `aud-do` already run into this issue? 

---

Also, I need to make a correct to my initial post. It turns out Nyquist process-type effects *can* call the aud-do interface. It's just that their I/O tracks and buffers are bound before they get any chance to aud-do (i.e. before "user/plugin code" executes), so the results [are not](https://github.com/audacity/audacity/issues/2464#issuecomment-1019348083) what one might expect.

**Comment by Paul-Licameli** [comment]
> > Giving Nyquist a foreign function interface that might reenter C++ world, but can catch C++ exceptions, unwind the C stack levels (and the Lisp call stack levels implemented thereby), then rethrow the exception into C++ again -- well, all of that would require care.
> 
> Doesn't `aud-do` already run into this issue?
> 

Yes, but it doesn't fix it yet.  Maybe you can amuse yourself figuring out how to crash the program this way, then open an issue.

You need to envision three layers, the C++ application, the Nyquist library in C, and the Lisp interpreter.  Each layer on that list can directly call the later ones.  That is the link-time dependency hierarchy, but the later layers can call back up through function pointers passed to them.  So the old C libraries can call newer C++.

`std::exception_ptr` would have to be part of the solution, to save an exception, return an error code, pop some C stack back to C++, detect the error, then rethrow through C++ stack levels.  C code can't to that, so all places that might call back into C++ should have some gate-keeping like that on the C++ side, or a proof that exceptions can't happen.

You can see limited use of that technique in Nyquist, but only because the appending to a WaveTrack to hold results of Nyquist evaluation might throw because of device space exhaustion or similar.

See `NyquistEffect::GetCallback` which uses `current_exception`. This is only lazy evaluation of the Nyquist sound object, which was previously prepared by a call into the XLisp interpreter which has returned already.  (But the lazy evaluation of some sounds can also re-invoke the interpreter!  See uses of `xleval` in Nyquist.)

See also `create_desc` which allows XLisp to garbage-collect foreign objects and the calls to that in Nyquist, to register its sound type as another primitive opaque type in XLisp.  Another dependency inversion, more "foreign" code for XLisp, supplied by the intermediate Nyquist level above it.  But no C++ layer there.

See the very bottom of Nyquist.cpp where Audacity binds some functions to names in the interpreter, giving it foreign functions that jump way up into C++ application land.

**Comment by werame** [comment]
> But the lazy evaluation of some sounds can also re-invoke the interpreter! 

Yeah, `seq` for instance (also `trigger` and another one I don't recall right now.)
